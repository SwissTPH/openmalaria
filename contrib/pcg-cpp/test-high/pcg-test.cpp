/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
