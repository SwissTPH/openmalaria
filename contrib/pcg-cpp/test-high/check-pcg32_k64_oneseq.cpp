#define RNG pcg32_k64_oneseq
#define TWO_ARG_INIT 0

#include "pcg-test.cpp"

