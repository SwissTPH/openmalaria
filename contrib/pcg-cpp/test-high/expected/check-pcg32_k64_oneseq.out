pcg32_k64_oneseq:
      -  result:      32-bit unsigned int
      -  period:      2^2110
      -  size:        264 bytes

Round 1:
  32bit: 0x8bfad379 0x21d323ad 0xd58574f9 0x282d3aa0 0x33b49afc 0x5821fdb2
  Again: 0x8bfad379 0x21d323ad 0xd58574f9 0x282d3aa0 0x33b49afc 0x5821fdb2
  Coins: TTHHTTHHTTTHTTTHTTHHTHHHHTTTTTTTHTHTHTHTHTTHHTHHTTHHTTHHHTTHHHTTH
  Rolls: 1 6 3 2 1 3 4 4 4 2 3 2 4 3 2 3 3 1 2 2 4 5 4 6 1 4 1 4 4 3 3 2 2
   -->   rolling dice used 33 random numbers
  Cards: 8s 5s 2d Qh Ad Kd 4c 3c 4d Ac Jd Js Kh 2h Qc Th Kc 6h 5d 3h Ah Td
	 7h 8c 9s 7s 9h Ks 4s 3d 2s 7c 7d 3s 5c 2c 6s 8d Jc 6d 9d Ts 5h Tc
	 6c Jh Qd 9c 4h 8h As Qs

Round 2:
  32bit: 0x97d54bf1 0xb85353db 0xc986820d 0x44c61d24 0x8263654e 0xcafc64df
  Again: 0x97d54bf1 0xb85353db 0xc986820d 0x44c61d24 0x8263654e 0xcafc64df
  Coins: TTHHHHHTTHTHHHHTTTHHHTHTTHTHTTHTHHTTHTHHTHHHTHTHHTTHTHTHHTHTTHHTH
  Rolls: 6 5 3 2 1 3 5 4 3 2 4 2 6 5 2 1 4 6 4 4 3 6 2 6 3 1 2 5 1 4 1 6 1
   -->   rolling dice used 33 random numbers
  Cards: Kc 9c Td 7d 4h 8s 8d 2h Ad Qh 4c 3s 6s Jh 8h Qs 7c 7h Th 9d 3c Js
	 2s 4d Kd 2c Tc As 4s 5d Ah 9s 5h 7s 9h Jd Ac 3h 2d Jc Ks Kh Qc Ts
	 3d 5c 6c 6d 5s 8c 6h Qd

Round 3:
  32bit: 0x2c1b99e5 0xf187ff1f 0xd1079e86 0x74c45a92 0x4d0b42be 0x652cd137
  Again: 0x2c1b99e5 0xf187ff1f 0xd1079e86 0x74c45a92 0x4d0b42be 0x652cd137
  Coins: HTHHTHTHTHTHTHHTTHHHHHHHTHHHHTHHTTHHTTTTHTTTHHTHTTHTTTTHHHHTHTHHH
  Rolls: 5 4 5 2 2 4 1 4 6 6 2 4 2 4 3 5 6 1 4 6 5 1 6 4 2 5 3 3 4 3 4 1 4
   -->   rolling dice used 33 random numbers
  Cards: 8s 8h 7s As Ac Qh 5h 3d Jd Ks 5s 4c Qc Td 4s 4h 3h Kd 2d 9d Tc 9c
	 8c 2s 6c Kc 7d 4d 9s Qs 6d 2h 8d 7h Ts Jh 7c 3c 5c Ad Ah 9h 6s Kh
	 Qd Th Js Jc 5d 3s 2c 6h

Round 4:
  32bit: 0x0166955b 0x98c26e63 0xa2eb39be 0x8e7a0f18 0xa1ee6dc4 0xa28d8093
  Again: 0x0166955b 0x98c26e63 0xa2eb39be 0x8e7a0f18 0xa1ee6dc4 0xa28d8093
  Coins: TTHHTTTTHHHHHHTHTHTTTTHTHTHTTTTHHHTTTTHTTHHTHHHTTTHHTTHTTHHHHHHHH
  Rolls: 5 4 1 1 3 3 3 2 3 2 1 2 1 2 5 6 6 4 6 5 5 3 6 6 4 2 2 6 1 1 4 1 1
   -->   rolling dice used 33 random numbers
  Cards: 8h Qc 7c 2d 2s 2c 8c 4d 7h Js 9c Td 5h 4c 7d 6c Ad Qh Kh 2h 8d Th
	 9d 6h 3s 6s 5c 5s 4s 3d Jc 3h 3c Jd 9h 9s Kc Kd Ks 4h Ah 8s Ts Tc
	 Qs 6d Ac 5d Jh As 7s Qd

Round 5:
  32bit: 0xc0f40efc 0x77437fab 0x80c937c8 0xc6ae4e8b 0xa9809bbb 0xe6523fd7
  Again: 0xc0f40efc 0x77437fab 0x80c937c8 0xc6ae4e8b 0xa9809bbb 0xe6523fd7
  Coins: HHHTTTHHTTTHHHHTTTTTHTTHTHHTHHHHHHTTTHTHTHHTTHTTHTTHHHHTTHHHTHHTH
  Rolls: 4 1 5 1 1 3 2 6 1 3 3 3 1 4 2 1 2 5 4 2 1 5 6 5 2 5 3 3 3 3 6 2 3
   -->   rolling dice used 33 random numbers
  Cards: 4d 9c 9s 8s Jc 7d 3c Ac 3d 5h 5c Js 2c Qs Th 4s 7h 2s 8c 4h Qd Tc
	 Ah 9h 3h Qh Jh 6s Jd 3s 8h Ts Kh Kd 4c 7s 2h 6c Qc 6d 8d Kc Ks Td
	 Ad As 5s 7c 6h 9d 5d 2d

